### PR TITLE
Resolve race in accessing/updating attribute maps

### PR DESCRIPTION
**Pull request by jeffreylovitz**
This PR adds race condition handling for the attribute mappings, solving any potential issues in our stress test and in general.

It must accommodate a few cases:
- Trying to look up an attribute from a reader query, such as the inner `_AR_EXP_UpdatePropIdx` call in `MATCH (x) RETURN x.name`.
- Adding a new attribute key from a non-locked area, such as in `PropertyMap_New` while converting an AST's CREATE clause.
- Adding a new attribute key from a locked region, such as committing an update in the critical region of `OpUpdate`.
- Adding a new attribute key from a non-Cypher endpoint like bulk insertion.

So far as I can tell, the changes here comprehensively handle all scenarios.

### CODE REVIEW

File: src/query_ctx.c [hunk 1]
```diff
@@ -171,6 +179,7 @@ bool QueryCtx_LockForCommit(void) {
 														  strlen(gc->graph_name));
 	char *error;
 	_QueryCtx_ThreadSafeContextLock(ctx);
+	ctx->internal_exec_ctx.locked_for_commit = true;
```
Comment (swilly22) [inline src/query_ctx.c:182]: Why was this line moved?
if `locked_for_commit` indicates that the graph is locked, (we're OK to modify the graph) then please revert this change.

File: src/graph/graphcontext.c [hunk 2]
```diff
@@ -250,24 +250,35 @@ uint GraphContext_AttributeCount(GraphContext *gc) {
 }
 
 Attribute_ID GraphContext_FindOrAddAttribute(GraphContext *gc, const char *attribute) {
+	// If we are already in a critical region, no further locking should be done in this function.
+	bool in_critical_region = QueryCtx_IsLocked();
```
Comment (swilly22) [inline src/graph/graphcontext.c:254]: Can't we simply try to reacquire the lock, `Graph_AcquireReadLock`
basically changing the lock to be reentrant.
if we'll do that this function and check can be removed.

File: src/graph/graphcontext.c [hunk 3]
```diff
@@ -250,24 +250,35 @@ uint GraphContext_AttributeCount(GraphContext *gc) {
...
+
+	/* If we are not already in a critical region, acquire a read lock.
+	 * Reader threads may acquire a read lock multiple times as long
+	 * as they release the lock the same number of times.
+	 * Writer threads that have not yet entered the critical region will hold no lock at this point,
+	 * while writer threads within the critical region are guaranteed to be the only running thread. */
```
Comment (swilly22) [inline src/graph/graphcontext.c:260]: I believe writer threads acquire a read lock in the same way read threads do, only difference is that writer threads will acquire write lock eventually (when performing the actual modification)

File: src/graph/graph.h [hunk 4]
```diff
@@ -87,15 +87,21 @@ void Graph_AcquireReadLock(Graph *g);
...
 void Graph_AcquireWriteLock(Graph *g);
 
+/* Release a read lock and acquire a write lock without modifying the '_writelocked' flag. */
+void Graph_UpgradeTemporaryLock(Graph *g);
+
+/* Release the held lock */
+void Graph_ReleaseLock(Graph *g);
+
+/* Release the held lock without modifying the '_writelocked' flag.  */
+void Graph_ReleaseTemporaryLock(Graph *g);
```
Comment (swilly22) [inline src/graph/graph.h:97]: Our locking logic is complex enough without these additions, can we use the graph RWLock without these additions?
I feel like they make this entire process more complex, if the answer is no, then I would prefer introducing a new mutex guarding the attribute rax and avoid mixing the graph lock with the attribute storage

File: src/graph/graphcontext.c [hunk 5]
```diff
@@ -276,9 +287,18 @@ const char *GraphContext_GetAttributeString(const GraphContext *gc, Attribute_ID
...
 Attribute_ID GraphContext_GetAttributeID(const GraphContext *gc, const char *attribute) {
-	Attribute_ID *id = raxFind(gc->attributes, (unsigned char *)attribute, strlen(attribute));
+	// If we are already in a critical region, no further locking should be done in this function.
+	bool in_critical_region = QueryCtx_IsLocked();
+	// Otherwise, acquire a read lock to ensure the attribute map is not being modified.
+	if(!in_critical_region) Graph_AcquireReadLock(gc->g);
+	// Look up the attribute ID.
+	void *id = raxFind(gc->attributes, (unsigned char *)attribute, strlen(attribute));
+	// Release the lock if acquired.
+	if(!in_critical_region) Graph_ReleaseTemporaryLock(gc->g);
```
Comment (swilly22) [inline src/graph/graphcontext.c:297]: I think it would be cleaner and less confusing / intertwine if we have dedicated RWLock specifically for `gc->attributes` mixing in the Graph lock make the system more complex, let's keep it simple.

File: src/graph/graphcontext.c [hunk 6]
```diff
@@ -276,9 +287,18 @@ const char *GraphContext_GetAttributeString(const GraphContext *gc, Attribute_ID
 }
 
 Attribute_ID GraphContext_GetAttributeID(const GraphContext *gc, const char *attribute) {
-	Attribute_ID *id = raxFind(gc->attributes, (unsigned char *)attribute, strlen(attribute));
+	// If we are already in a critical region, no further locking should be done in this function.
+	bool in_critical_region = QueryCtx_IsLocked();
```
Comment (swilly22) [inline src/graph/graphcontext.c:291]: I hope this function isn't called multiple times for a given `attribute` during the lifetime of a query

File: src/graph/graphcontext.c [hunk 7]
```diff
@@ -250,35 +252,50 @@ uint GraphContext_AttributeCount(GraphContext *gc) {
...
-				  NULL);
-		gc->string_mapping = array_append(gc->string_mapping, rm_strdup(attribute));
+	void *attribute_id = raxFind(gc->attributes, (unsigned char *)attribute, strlen(attribute));
+
+	if(attribute_id == raxNotFound) {
+		// We are writing to the shared GraphContext; release the held lock and re-acquire as a writer.
+		pthread_rwlock_unlock(&gc->_attribute_rwlock);
+		pthread_rwlock_wrlock(&gc->_attribute_rwlock);
+
+		attribute_id = (void *)raxSize(gc->attributes);
```
Comment (swilly22) [inline src/graph/graphcontext.c:266]: This is problematic, consider 2 threads all trying to introduce the same set of attributes: ('a', 'b', 'c')
if thread 1 managed to acquire the lock and insert all 3 attributes, while thread 2 is trying to insert attribute 'a'  once 2 gets the lock (trying to introduce 'a') rax size is 3 but the attribute id for 'a' is 1.

File: src/graph/graphcontext.c [hunk 8]
```diff
@@ -250,35 +252,50 @@ uint GraphContext_AttributeCount(GraphContext *gc) {
...
+		/* rc will be 1 if the insertion was successful and 0 if the rax was not updated,
+		 * which could occur if 2 threads try to retrieve the same attribute simultaneously.
+		 * In this case the second thread will not modify the rax or the string mapping. */
+		int rc = raxTryInsert(gc->attributes,
```
Comment (swilly22) [inline src/graph/graphcontext.c:270]: `raxTryInsert` will not update if the key exists ?

File: src/ast/ast_shared.c [hunk 9]
```diff
@@ -106,8 +106,7 @@ void PropertyMap_Free(PropertyMap *map) {
 EntityUpdateEvalCtx EntityUpdateEvalCtx_Clone(EntityUpdateEvalCtx ctx) {
 	EntityUpdateEvalCtx clone;
```
Comment (swilly22) [inline src/ast/ast_shared.c:107]: Might be more efficient to use memcpy and `AR_EXP_Clone`

File: src/ast/ast_build_op_contexts.c [hunk 10]
```diff
@@ -66,24 +66,25 @@ static EntityUpdateEvalCtx _NewUpdateCtx(const cypher_astnode_t *set_item) {
 	/* Track all required information to perform an update. */
 	const char *alias = entity->operand.variadic.entity_alias;
 	const char *attribute = cypher_ast_prop_name_get_value(prop);
+	Attribute_ID attribute_id = GraphContext_FindOrAddAttribute(gc, attribute);
```
Comment (swilly22) [inline src/ast/ast_build_op_contexts.c:69]: Should we assert making sure `attribute_id` exists ?

File: src/ast/ast_build_op_contexts.c [hunk 11]
```diff
@@ -157,8 +158,8 @@ void AST_PreparePathCreation(const cypher_astnode_t *path, QueryGraph *qg, rax *
 	}
 }
 
-AST_MergeContext AST_PrepareMergeOp(const cypher_astnode_t *merge_clause, QueryGraph *qg,
-									rax *bound_vars) {
+AST_MergeContext AST_PrepareMergeOp(const cypher_astnode_t *merge_clause, GraphContext *gc,
```
Comment (swilly22) [inline src/ast/ast_build_op_contexts.c:161]: Can't we get `QueryGraph` from `GraphContext` ?

File: src/arithmetic/arithmetic_expression.c [hunk 12]
```diff
@@ -22,6 +22,9 @@
 #include <ctype.h>
 #include <assert.h>
 
+// Property keys in variadic expressions will be ATTRIBUTE_UNSET until the first lookup.
+#define ATTRIBUTE_UNSET (USHRT_MAX - 1)
```
Comment (swilly22) [inline src/arithmetic/arithmetic_expression.c:26]: #define ATTRIBUTE_UNSET (ATTRIBUTE_NOTFOUND - 1)

File: src/ast/ast_build_op_contexts.c (hunk 11 above)
Comment (jeffreylovitz) [reply src/ast/ast_build_op_contexts.c:161]: No, the `QueryGraph` lives on the `ExecutionPlan`.
